---
title: "scRNA-seq quality control"
format: html
---

Load `Seurat`.

```{r load_seurat}
library("Seurat")
packageVersion("Seurat")
```

Read HDF5 file.

```{r read10x_h5}
bone <- Read10X_h5('../data/sc_best_practices/filtered_feature_bc_matrix.h5')
str(bone)
```

Create Seurat object. Matches dimension from [Best Practices guide](https://www.sc-best-practices.org/preprocessing_visualization/quality_control.html#environment-setup-and-data) (AnnData object with n_obs × n_vars = 16934 × 36601).

```{r seurat_obj}
seurat_obj <- CreateSeuratObject(
  counts = bone$`Gene Expression`,
  project = 'bone marrow'
)

seurat_obj
```

Median absolute deviations (MAD).

$$
MAD = median(|X_i - median(X)|)
$$

Compare the `mad()` function with a plain implementation. Note that `mad()` adjust by a factor for asymptotically normal consistency by default; set `constant = 1` to get the same results.

```{r compare_mad}
set.seed(1984)
x <- sample(x = 1:100, size = 1000, replace = TRUE)
median(abs(x - median(x)))
mad(x, constant = 1)
```
